Escort as top level object

lostie opened the pull request:
Refactor data modelling to guarantee idempotency of escort

As it stands every time a new escort was created the information used
for a previous escort, namely, detainee information, risk assessment,
healthcare assessment, offences would be re-used and most likely
modified which meant we would track of the information used to perform
the previous escort.

These changes ensure that all the information used in an escort are
related only to that escort and cannot be modify outside the scope of
the escort itself.

This turns out to be quite a big change in terms of:
- Routes - all the data required to be created/modified is now under the
escorts scope
- Detainee and Move are not longer directly related but instead they
    belong to an Escort record. There's potential, in the future for these
    two models to be normalised and all the information related with them to
    be present directly in the Escort record. Due to the nature of the
    refactor we have refrained from doing any more changes than the MVP.
- The risk and healthcare assessments and the offences are currently
    related to a detainee, so the escort record access its information by
    delegating their call to the detainee. There's potential for this to be
    move directly to the escort, so that the escort has a directly
    relationship with the data it requires
- The workflows associated with the move, the risk and healthcare
    assessments, and the offences, are defined in a very convoluted way
    where the detainee holds the relation to the data but the move holds the
    relation to the status of the data. In this PR we masked this by having
    the models themselves to response to status enquires. Under the hood, we
    are still navigation to this convoluted way of accessing the status. In
    a future iteration we will simplify the process by having the model to
    respond directly to this status enquires.
- This PR starts to establish the responsability to the status of the
    escort to the escort itself, although it currently does it by delegating
    some of its status checks to the Move record. This is a transiction step
    as the Move record has a lot of responsability that shouldn't really
    have, but again, refactoring this would increase even more the size of
    the PR which is already quite big.

Future iterations:
- Establish correct relationships between models
- Have escort to control directly its status

Commit:
Refactor data modelling to guarantee idempotency of escort

As it standed everytime a new escort was created the information used
for a previous escort, namely, detainee information, risk assessment,
healthcare assessment, offences would be re-used and most likely
modified which meant we would track of the information used to perform
the previous escort.

These changes ensure that all the information used in an escort are
related only to that escort and cannot be modify outside the scope of
the escort itself.

This turns out to be quite a big change in terms of:
- Routes - all the data required to be created/modified is now under the
escorts scope
- Detainee and Move are not longer directly related but instead they
belong to an Escort record. There's potential, in the future for these
two models to be normalised and all the information related with them to
be present directly in the Escort record. Due to the nature of the
refactor we have refrained from doing any more changes than the MVP.
- The risk and healthcare assessments and the offences are currently
related to a detainee, so the escort record access its information by
delegating their call to the detainee. There's potential for this to be
move directly to the escort, so that the escort has a directly
relationship with the data it requires
- The workflows associated with the move, the risk and healthcare
assessments, and the offences, are defined in a very convoluted way
where the detainee holds the relation to the data but the move holds the
relation to the status of the data. In this PR we masked this by having
the models themselves to response to status enquires. Under the hood, we
are still navigation to this convoluted way of accessing the status. In
a future iteration we will simplify the process by having the model to
respond directly to this status enquires.
- This PR starts to establish the responsability to the status of the
escort to the escort itself, although it currently does it by delegating
some of its status checks to the Move record. This is a transiction step
as the Move record has a lot of responsability that shouldn't really
have, but again, refactoring this would increase even more the size of
the PR which is already quite big.

Future iterations:
- Establish correct relationships between models
- Have escort to control directly its status

lostie commented on app/views/escorts/show.html.slim:
```diff
@@ -0,0 +1,21 @@
+    - breadcrumb detainee_breadcrumb(escort.detainee), '#'
+
+.escort
+  = render partial: 'header', locals: { detainee: escort.detainee, alerts: DetaineeProfileAlertsPresenter.new(escort.detainee) }
```
We could rename this to `EscortAlertsPresenter`

dan-palmer commented on db/migrate/20170405094953_re_create_escorts.rb:
```diff
@@ -0,0 +1,9 @@
+class ReCreateEscorts < ActiveRecord::Migration[5.0]
+  def change
+    create_table :escorts, id: :uuid do |t|
+      t.string :prison_number
```
Should we remove the prison number from the detainee table and just delegate to the escorts prison number for now?

dan-palmer commented on app/services/escort_creator.rb:
```diff
@@ -0,0 +1,43 @@
+      :risk,
+      { healthcare: [:medications] },
+      { offences: [:current_offences] }
+] },
```
🚓 such padding

dan-palmer commented on app/controllers/application_controller.rb:
```diff
@@ -5,23 +5,11 @@ class ApplicationController < ActionController::Base
 
   before_action :authenticate_user!
 
-  helper_method :offences,
```
👍

dan-palmer commented on app/views/homepage/_no_result.html.slim:
```diff
@@ -1,6 +1,9 @@
 p
-  | There is no previously created profile on
+  | There is no previously PER record on
```
Probably should read `There are no previously created PER records on` ..

dan-palmer commented on spec/features/search_spec.rb:
```diff
@@ -61,16 +67,17 @@ def expect_error_message
 
   def expect_result_with_no_move(detainee)
-    expect(page).to have_link('Add new move', href: new_detainee_move_path(detainee))
+    # expect(page).to have_button('Add new move', href: new_escort_move_path(detainee.escort))
```
♻️

dan-palmer commented on spec/services/create_escort_spec.rb:
```diff
@@ -0,0 +1,116 @@
+
+    expected_current_offences_attributes = existent_escort.offences.current_offences.map { |co| co.attributes.except(*except_current_offences_attributes) }
+    current_offences_attributes = new_escort.offences.current_offences.map { |co| co.attributes.except(*except_current_offences_attributes) }
+    expected_current_offences_attributes.each do |expected_attributes|
```
Given the hashes have the same schemas on the lhs and rhs of the test would `match_array` not work here? It doesn't take into account the order of the arrays just that they contain the same things.

```ruby
expect(current_offences_attributes).to match_array(expected_current_offences_attributes)
```

lostie replied:
That does not work as expected:

```ruby
     Failure/Error: current_offences_attributes.to match_array(expected_current_offences_attributes)

     NoMethodError:
       undefined method `>=' for #<RSpec::Matchers::BuiltIn::ContainExactly:0x007fb4c6ceff68>
     # ./spec/services/create_escort_spec.rb:70:in `expect_offences_to_be_cloned'
     # ./spec/services/create_escort_spec.rb:31:in `block (3 levels) in <top (required)>'
```

The current test runs as expected, so leaving it as is for now. Can be re-visited if some other way is found to achieve the same expectation.

dan-palmer replied:
Looks like a typo. The `current_offences_attributes` needs to be passed to the expect method in order to return an object that responds to `to`.

Commit:
Refactor dashboard to rely on escort rather than move

Escort is now the top level object and all the search results presented
in the dashboard derive from it.

Commit:
Further refactor to simplify escort creation

- Detainees controller no longer requires a prison number to be provided
as its under the scope of an escort which requires that information to
be provided.
- Re-worded the buttons to create/continue a new escort as the current
wording was quite confusing having the term 'profile' although there's
no such thing in the application
- Remove some of the indirection present in the view, so that at the
most the models call the information their require directly.

Commit:
Minor padding, re-wording and spec design fixes

lostie commented:
♻️

dan-palmer commented:
<img src="https://media.giphy.com/media/Hw8vYF4DNRCKY/giphy.gif" />
